lsbfdump-jni-rs/native/src/helper_functions.rs

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// View https://rust-lang.github.io/api-guidelines/naming.html for naming conventions for functions
+
+use crate::stdin_seek::StdinWithSeek;
+
+use std::fs::File;
+use std::io::{self, Read, Seek, SeekFrom};
+
+
+pub fn lsbf_dump(reader: &mut dyn Read, offset: u64, length: u64, show_address: bool) -> io::Result<Vec<String>> {
+  let mut lines = Vec::new();
+  let mut count = 0;
+  let mut buffer = [0; 4];
+
+  while count < length {
+    match reader.read_exact(&mut buffer) {
+      Ok(_) => {
+        let byte_strings = buffer.iter().copied().rev().map(byte_to_binary_string).collect::<Vec<String>>().join(" ");
+        let extra_spaces = if buffer.len() == 4 { "".to_string() } else { "         ".repeat(4 - buffer.len()) };
+        let address_string = if show_address { format!(" | 0x{:08X}", count + offset) } else { "".to_string() };
+        lines.push(format!("{}{}{}", extra_spaces, byte_strings, address_string));
+        count += 4;
+      }
+      Err(_) => break,
+    }
+  }
+
+  Ok(lines)
+}
+
+pub fn open_and_seek_input_stream(filename: &str, offset: u64) -> io::Result<Box<dyn Read>> {
+  if filename == "-" {
+    let mut stdin_with_seek = StdinWithSeek;
```

Comment:
You could just call read in a loop N times to carry out the seeking right here. I don't see the need for StdinWithSeek at all. See comment later about buffering vs. streaming behavior also. 

If you do that, then I don't see a reason for a dyn anything here. It would just be a `Box<Read>` wouldn't it?

Reply:
So the `dyn` is needed even after updating the code to loop over read. Whenever I would try to remove it, I would get an error `add 'dyn' keyword before this trait: 'dyn'`. I found online it is required, below is a quote from an answer found at https://users.rust-lang.org/t/why-does-rustc-not-automatically-infer-dyn-for-trait-objects/98535:

```
The dyn was added, and later made mandatory in the 2021 edition 1 because otherwise
it is very hard for a programmer reading the code or its documentation to know what is a
concrete type and what is a trait denoting a trait object. Being explicit about things like this
is an important part of Rust's style.
```

Reply:
So dyn is required if you are handling an object as an instance of a Trait or base class vs. as the actual class object.

Ok. Seems to fly in the face of OOP principles, but Ok.